coderd/workspaceapps.go

```diff
@@ -231,22 +268,159 @@ func (api *API) parseWorkspaceApplicationHostname(rw http.ResponseWriter, r *htt
+// agent and returns it. If the application is not found or there was a server
+// error while looking it up, an HTML error page is returned and false is
+// returned so the caller can return early.
+func (api *API) lookupWorkspaceApp(rw http.ResponseWriter, r *http.Request, agentID uuid.UUID, appName string) (database.WorkspaceApp, bool) {
+	app, err := api.Database.GetWorkspaceAppByAgentIDAndName(r.Context(), database.GetWorkspaceAppByAgentIDAndNameParams{
+		AgentID: agentID,
+		Name:    appName,
+	})
+	if xerrors.Is(err, sql.ErrNoRows) {
+		renderApplicationNotFound(rw, r, api.AccessURL)
+		return database.WorkspaceApp{}, false
+	}
+	if err != nil {
+		site.RenderStaticErrorPage(rw, r, site.ErrorPageData{
+			Status:       http.StatusInternalServerError,
+			Title:        "Internal Server Error",
+			Description:  "Could not fetch workspace application: " + err.Error(),
+			RetryEnabled: true,
+			DashboardURL: api.AccessURL.String(),
+		})
+		return database.WorkspaceApp{}, false
+	}
+
+	return app, true
+}
+
+func (api *API) authorizeWorkspaceApp(r *http.Request, sharingLevel database.AppSharingLevel, workspace database.Workspace) (bool, error) {
+	ctx := r.Context()
+
+	// Short circuit if not authenticated.
+	roles, ok := httpmw.UserAuthorizationOptional(r)
+	if !ok {
+		// The user is not authenticated, so they can only access the app if it
+		// is public.
+		return sharingLevel == database.AppSharingLevelPublic, nil
+	}
+
+	// Do a standard RBAC check. This accounts for share level "owner" and any
+	// other RBAC rules that may be in place.
+	//
+	// Regardless of share level or whether it's enabled or not, the owner of
+	// the workspace can always access applications (as long as their key's
+	// scope allows it).
+	err := api.Authorizer.ByRoleName(ctx, roles.ID.String(), roles.Roles, roles.Scope.ToRBAC(), []string{}, rbac.ActionCreate, workspace.ApplicationConnectRBAC())
+	if err == nil {
+		return true, nil
+	}
+
+	switch sharingLevel {
+	case database.AppSharingLevelOwner:
+		// We essentially already did this above.
+	case database.AppSharingLevelTemplate:
+		// Check if the user has access to the same template as the workspace.
+		template, err := api.Database.GetTemplateByID(ctx, workspace.TemplateID)
+		if err != nil {
+			return false, xerrors.Errorf("get template %q: %w", workspace.TemplateID, err)
+		}
+
+		// We have to perform this check without scopes enabled because
+		// otherwise this check will always fail on a scoped API key.
+		err = api.Authorizer.ByRoleName(ctx, roles.ID.String(), roles.Roles, rbac.ScopeAll, []string{}, rbac.ActionRead, template.RBACObject())
+		if err != nil {
+			// Exit early if the user doesn't have access to the template.
+			return false, nil
+		}
+
+		// Now check if the user has ApplicationConnect access to their own
+		// workspaces.
+		object := rbac.ResourceWorkspaceApplicationConnect.WithOwner(roles.ID.String())
```

Comment:
We should add the org too for template scope.

```suggestion
		object := rbac.ResourceWorkspaceApplicationConnect.WithOwner(roles.ID.String()).InOrg(workspace.OrganizationID)
```

Reply:
template level has been removed